### ISSUE TITLE
Is it correct to cache this per-fiber?

### ISSUE THREAD

**Issue by ioquatix** [mperham/connection_pool#168]
https://github.com/mperham/connection_pool/blob/428c06f34209ee7a99f1ace5b96e567841c00d1c/lib/connection_pool.rb#L117

Looking at this code, I noticed it caches a connection per fiber.

I'm a little concerned that someone might expect:

```ruby
pool.with |connection1|
  connection1 ... # start a pipeline
  pool.with |connection2|
    connection2.get(...) # get some value
  end
  connection1 ... # end/commit pipeline
end
```

I don't understand why `connection1` and `connection2` should be the same connection. I would probably expect them to be distinct, otherwise the state from one might leak into another. It's probably even more odd if this operation is inside a library, that might be expecting a clean connection, but is instead gets spooky action at a distance with a user connection.

**Comment by mperham** [comment]
Yes this is intentional: while a connection is checked out, further checkouts will return the same connection. It's had this semantic for many years now.

**Comment by ioquatix** [comment]
Based on your response, there are two things I think we can do:

- Change the behaviour so that the same connection is not returned. Because a connection pool is a cache, I don't think anyone should be depending on this behaviour, i.e. a connection pool could be validly implemented by just returning a completely new connection every time (but would be slow), OR
- Add documentation regarding the situations in which `with do |connection|` can actually return a connection which is already in use and may have state (e.g. transactions, pipeline, etc) attached to it.

**Comment by ms-jpq** [comment]
wait, what is the recommended approach to avoid this issue, my use case involves building a library that can maintain both long lived HTTP connections, and short lived http requests.

with this connection sharing issue, it means that the long lived http connection and short lived http requests can stumble over each other if they are intermixed correct?

I will have no control over how the users would want to call my code, so this might be slightly tricky

**Comment by ioquatix** [comment]
The only recommended approach I can suggest right now is to avoid this library if this behaviour is problematic for your use case.

**Comment by mperham** [comment]
I'm not sure what to do with this issue. Close it? Provide a new API with different semantics?

**Comment by ioquatix** [comment]
I would suggest:

1. Add documentation regarding the situations in which with do |connection| can actually return a connection which is already in use and may have state (e.g. transactions, pipeline, etc) attached to it.
2. Add an option to control whether the same connection is returned.

Possible options for 2:

- Add an option to the pool, e.g. `ConnectionPool.new(shared_with: true/false)`. Keep the current default, but start issuing a warning. At some point in the future, change the default if otherwise unspecified.
- Add an option to `#with`, e.g. `#with(shared: false) do ...` that can't reuse any existing connection associated with the thread. Keep the current default, but start issuing a warning. At some point in the future, change the default if otherwise unspecified.
  - Alternatively, introduce a new method, e.g. `#checkout do ...` or `#acquire do ...` which have the new semantics of not sharing.

**Comment by ms-jpq** [comment]
My interim work around will probably just to 

1) still use the library, but on every `with(&blk)` from the pool, check if the current Thread / Fiber has been marked with a Fiber / Thread local

2) If a mark has been found, I will just skip pulling from the connection pool, and create a new session instead.

3) Otherwise mark the current Fiber / Thread, and pull from the pool.

---

I do wonder if for step 2), I can just wrap the `&blk` with a `Fiber.new { blk.call }.resume` and thus avoid creating a session from scratch.

My main worry is that this would interact poorly with either the Fiber schedulers, or some other use case of Fibers that I am not aware of.

**Comment by mperham** [comment]
I don't think it's a good idea for a single Thread to juggle multiple connections at the same time from the same pool. That's more for an event reactor and this library is not meant for that pattern. That why it contains this checkout semantic; I think that's the impedance mismatch which you're running up against.

Let me be clear: I don't want connection_pool to actively work against the async pattern but I also want to recognize that connection_pool is not meant to be colorless (i.e. to support both blocking and non-blocking usage). I already maintain a large amount of Ruby infrastructure and I don't have the brain power to become an expert in async in order to support connection_pool with async usage patterns. That's a space I choose to ignore because it contains a lot of "mental model" complexity. I hope that makes sense.

If there are small changes or a simple API we can add to make things easier, please open a PR. I would be happy to get something (especially from @ioquatix as the undisputed expert in the field) which makes async usage safer or easier. If a Thread can checkout multiple connections, I worry about connection exhaustion bugs; a safer, block-oriented API would be much preferred.

BTW https://github.com/drbrain/net-http-persistent is a gem which uses connection_pool and provides long-lived connections. I presume it did not meet your needs?

**Comment by ioquatix** [comment]
I don’t think the concern is only about “a single thread juggling multiple connections.” The bigger issue is that two different code paths can unexpectedly share the **same connection**, even when the pool has more than one connection available.

`connection_pool` deliberately re-enters with the same connection within the same thread/fiber. That means nested `with` blocks don’t check out a fresh connection, they hand you the already-checked-out one. This can lead to surprising behaviour when the connection carries state.

For example, consider two libraries that don’t know about each other:

```ruby
$redis_pool = ConnectionPool.new(size: 5, timeout: 5) { Redis.new }

module LibraryA
  def self.do_something
    $redis_pool.with do |redis|
      redis.multi
      redis.set("foo", "bar")
      yield
      redis.exec
    end
  end
end

module LibraryB
  def self.lookup(key)
    $redis_pool.with do |redis|
      redis.get(key)
    end
  end
end

LibraryA.do_something do
  value = LibraryB.lookup("foo")
  puts "lookup returned: #{value.inspect}"
end
```

What actually happens:

1. Library A starts a transaction (`MULTI`) and queues `SET foo bar`.
2. The application calls Library B’s `lookup`.
3. Because we’re in the same fiber, `pool.with` reuses the very same connection that is still inside a transaction.
4. The `GET foo` is not executed immediately — it is queued into the open transaction. The client returns `nil` to Library B.
5. When Library A calls `EXEC`, Redis finally executes both commands (`SET` and `GET`). The `GET` result is discarded, since Library B already got back `nil`.

From the perspective of the caller, `LibraryB.lookup` looks like a simple read, but it was silently swallowed into Library A’s transaction. This is extremely surprising, and can corrupt transactional boundaries or pollute unrelated code paths.

So the problem isn’t just threads juggling multiple connections — it’s that any nested `with` can cause different logical operations to collide on the same connection. That’s why implicit sharing like this is dangerous.

**Comment by womblep** [comment]
I get what you are trying to say but I think that example is stretching what this library IMHO is used for. The same problem exists if you just got a redis connection and passed it to the 2 libraries so I am not sure that it is that surprising.

You could arguably have an example that was deeply nested and exhausted the pool in also surprising ways if it checked out a new connection each time.

A new method or class I think would solve the issue. It would also have the benefit of not breaking a lot of code that depends on the current behavior.

**Comment by mperham** [comment]
I'd suspect that many applications are relying on this connection reuse semantic, especially to minimize the number of connections created in low-cost systems where a SaaS service forces a very tight connection budget. Changing the semanic IMO is a major decision and worthy of a major version bump.

connection_pool has been heavily used in multi-threaded processes for a decade now and I can't recall anyone reporting an issue of the type you are describing. If you want a different semantic for Fibers, I'm ok with that since Fibers are far more likely to have multiple transactions in-flight on one Thread, PRs welcome there.
